### PR TITLE
WIP: pythonPackages: normalize attribute names, fixes #49691

### LINKED ISSUE #49691: Python package set: normalize all attribute names
## Issue description

The convention in `python-packages.nix` is to use [normalized names](https://www.python.org/dev/peps/pep-0503/#normalized-names) although far from every attribute fulfills this requirement.

Having all packages conform will simplify things for automation.

Note that the scope is only the attribute names, and not the `pname` attributes.

cc @costrouc

### PR DESCRIPTION
**Pull request by FRidh**
###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### To do

- [x] check attribute names against regex
- [ ] fix all attribute names (sigh...) https://github.com/NixOS/nixpkgs/issues/245383
- [ ] 0 rebuilds
- [ ] document this requirement (backwards incompatible!)
- [ ] optional: check also the pname attribute. That requires careful checking of hashes because pname is typically inherited by the fetchers.

### CODE REVIEW

File: pkgs/top-level/python-packages.nix [hunk 1]
```diff
@@ -7856,4 +7856,13 @@ in {
 
 });
 
-in fix' (extends overrides packages)
+# Require normalized attribute names as these are predictable.
+# https://www.python.org/dev/peps/pep-0503/#normalized-names
+#isNormalised = name: (builtins.match "^([A-Z0-9]|[A-Z0-9][A-Z0-9]*[A-Z0-9])$") name == null;
+isNormalised = name: name == requiredName name;
```
Comment (FRidh) [inline pkgs/top-level/python-packages.nix]: This is too strict. We now cannot have variants such as `pyqt5_with_qtwebkit` or `pytest_6`. Having these versions at the end is possibly to add with a regex, but variants such as `pyqt5_with_qtwebkit` are not possible to distinguish from non-normalized names.

Comment (FRidh) [reply pkgs/top-level/python-packages.nix]: Note evaluation is lazy, so you only get the throw when explicitly requesting a `pytest_6`. Still, it is a problem.

Comment (FRidh) [reply pkgs/top-level/python-packages.nix]: One could partially work-around this issue by requiring overriding the Python package set to use one of these variants. At least that works for applications. And libraries should not be able to use multiple variants anyway. The limitation is test dependencies, there you may be able to use different versions.

**Comment by jonringer** [comment]
also, it's common to prefix with "_" if the package name starts with a number.

maybe we could do `builtins.match` and have a regex check the format, and allow for version suffixes e.g. ( `(_\d+)?`)

File: pkgs/top-level/python-packages.nix (hunk 1 above)
Comment (FRidh) [reply pkgs/top-level/python-packages.nix]: @jonringer there is now a `toVariant` function for declaring when a derivation is a variant that can have a `_suffix`.

**Comment by FRidh** [comment]
Got a working regex now. I am by no means a regexpert so I welcome any improvements.

Right now, the following names are allowed:
- `numpy`
- `numpy_1`
- `numpy_1_2_30`
- `_2to3`
- `numpy-foo-bar_1_2_30`

Not allowed are:
- `numpy_foo`
- `numpy_foo_bar`
- `numpy_foo_bar_1`

It means variants need to be moved away from the top-level, to the `passthru` of the original package. That also means they typically will not be built by Hydra.

**Pull request closed without merging**